pandas/core/reshape/merge.py

```diff
@@ -2355,12 +2355,14 @@ def _factorize_keys(
     rk = extract_array(rk, extract_numpy=True, extract_range=True)
     # TODO: if either is a RangeIndex, we can likely factorize more efficiently?
 
-    if isinstance(lk.dtype, DatetimeTZDtype) and isinstance(rk.dtype, DatetimeTZDtype):
+    if (isinstance(lk, DatetimeArray) and isinstance(rk, DatetimeArray)) and (
+        lk._is_tzawareness_compat(rk)
```

Comment:
No idea, but that was also not the reason for using this: what we want to check here is that we have either two DatetimeTZDtypes, or two np.dtype("M8"). 

So initially I expanded the original `if isinstance(lk.dtype, DatetimeTZDtype) and isinstance(rk.dtype, DatetimeTZDtype)`  with something like `or (isinstance(lk.dtype, np.dtype) and lk.dtype.kind == "M" and isinstance(rk.dtype, np.dtype) and rk.dtype.kind == "M")`. But this got a bit long, and so this seemed simpler (and since we are calling a specific DatetimeArray method in the if-block, it also make sense to check for it). 

(the performance benefit of the PR comes from avoiding using the ObjectFactorizer)